packages/client/src/objectSet/ObjectSetListenerWebsocket.ts

```diff
@@ -72,74 +105,107 @@ export class ObjectSetListenerWebsocket {
   #lastWsConnect = 0;
   #client: MinimalClient;
 
-  /** map of listenerId to listener */
-  #listeners = new Map<
+  #logger?: Logger;
+
+  /**
+   * map of requestId to all active subscriptions at the time of the request
+   */
+  #pendingSubscriptions = new Map<
     string,
-    {
-      listener: ObjectSetListener<any>;
-      subscriptionId?: string;
-      objectSet: ObjectSet;
-      expiry: NodeJS.Timeout;
-    }
+    Subscription[]
   >();
 
-  /** map of subscriptionId to listenerId */
-  #subscriptionToRequestId = new Map<string, string>();
+  /**
+   * Map of subscriptionId to Subscription. Note: the subscriptionId may be
+   * temporary and not the actual subscriptionId from the server.
+   */
+  #subscriptions = new Map<
+    string,
+    Subscription
+  >();
 
-  #conjureContext: ConjureContext;
+  #oswContext: ConjureContext;
   #metadataContext: ConjureContext;
   #ossContext: ConjureContext;
 
+  #maybeDisconnectTimeout: ReturnType<typeof setTimeout> | undefined;
+
   private constructor(client: MinimalClient) {
     this.#client = client;
+    this.#logger = client.logger?.child({}, {
+      msgPrefix: "<OSW> ",
+    });
+    invariant(
+      client.stack.startsWith("https://") || client.stack.startsWith("http://"),
+      "Stack must be a URL",
+    );
 
-    const stackUrl = new URL(client.stack);
-    this.#conjureContext = {
-      baseUrl: stackUrl.origin,
+    this.#oswContext = {
+      baseUrl: client.stack,
       servicePath: "/object-set-watcher/api",
       fetchFn: client.fetch,
       tokenProvider: async () => await client.tokenProvider(),
     };
     this.#ossContext = {
-      baseUrl: stackUrl.origin,
+      ...this.#oswContext,
       servicePath: "/object-set-service/api",
-      fetchFn: client.fetch,
-      tokenProvider: async () => await client.tokenProvider(),
     };
     this.#metadataContext = {
-      baseUrl: stackUrl.origin,
+      ...this.#oswContext,
       servicePath: "/ontology-metadata/api",
-      fetchFn: client.fetch,
-      tokenProvider: async () => await client.tokenProvider(),
     };
   }
 
-  subscribe<Q extends ObjectOrInterfaceDefinition>(
+  async subscribe<Q extends ObjectOrInterfaceDefinition>(
     objectSet: ObjectSet,
     listener: ObjectSetListener<Q>,
-  ): () => void {
-    const requestId = crypto.randomUUID();
-    const expiry = setTimeout(() => {
-      this.#expire(requestId);
-    }, ONE_DAY_MS);
-    this.#listeners.set(requestId, { listener, objectSet, expiry });
-    this.#subscribe(requestId, objectSet);
+  ): Promise<() => void> {
+    globalThis.crypto ??= (await import("node:crypto")).webcrypto as any;
+    const sub: Subscription = {
+      listener: fillOutListener<any>(listener),
```

Comment:
why is this `any` vs `Q` ?

Reply:
Because it's going into a generic storage for all subscriptions. The type information is effectively erased at that point and the use of generics here is to make it easier for the consumer not us.

Reply:
comment?